test/test_c10d.py

```diff
@@ -3098,6 +3098,29 @@ def test_invalid_nccl_blocking_wait_env(self):
         self._run_invalid_nccl_blocking_wait_env('2147483647')
         self._run_invalid_nccl_blocking_wait_env('4294967295')
 
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_nccl_timeout(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        os.environ["NCCL_BLOCKING_WAIT"] = "1"
+
+        # Initialize process_group.
+        timeout = 1
+        c10d.distributed_c10d.init_process_group(
+            backend=dist.Backend.NCCL, store=store, world_size=2, rank=self.rank,
+            timeout=timedelta(seconds=timeout))
+        c10d.distributed_c10d.all_reduce(torch.rand(10).cuda(self.rank))
+
+        if self.rank == 0:
+            # This should timeout in about 1 second.
+            start = time.time()
+            with self.assertRaises(RuntimeError):
+                c10d.distributed_c10d.all_reduce(torch.rand(10).cuda(self.rank))
+
+            total_time = time.time() - start
+
+            self.assertLess(abs(total_time - timeout), 0.5)
```

Comment:
Why assert less than 0.5 if the timeout is 1 second?

To really test the timeout path (instead of the other process terminating), rank 1 should sleep until rank 0 has timed out.

Reply:
0.5 is just the leeway not the actual timeout. Ideally total_time - timeout should be 0, but there would be some overhead I'm accounting for and being conservative with 0.5 to avoid this test being flaky.